### PR TITLE
WDPD-98 Display paymentSDK error

### PR DESCRIPTION
**Pull request by ps-mstifter**
### This PR

* Displays errors returned from the paymentSDK in the payment step of the checkout process
* Displays error from check checkout form in javascript
* refactor response handling

### Jira Links

* https://jira.parkside.at/browse/WDPD-98

### CODE REVIEW

File: Extend/Controller/OrderController.php [hunk 1]
```diff
@@ -238,16 +239,34 @@ private function _handleResponse($oResponse, $oLogger, $oOrder)
...
+            $sSid = Registry::getSession()->sid(true);
+
+            $aQueryParams = [
+                'cl' => 'payment',
+                'lang' => $sBaseLanguage,
+            ];
+
+            $sFullQueryString = http_build_query($aQueryParams) . '&' . $sSid;
+
+            return Registry::getUtils()->redirect(Helper::getShopUrl() . 'index.php?' . $sFullQueryString);
```
Comment (ps-ohanel) [inline Extend/Controller/OrderController.php]: There is getShopHomeUrl() method on Config. Couldn't it be reused?

Comment (ps-mstifter) [reply Extend/Controller/OrderController.php]: Thanks for the tip, this reduced the code significantly :)

File: Extend/Controller/OrderController.php [hunk 2]
```diff
@@ -232,22 +232,37 @@ private function _handleTransaction($oBasket, $oOrder)
...
+
                 $oLogger->error("\t$sSeverity with code $sCode and message '$sDescription' occurred.");
             }
-            return;
+
+            // set the custom payment error code and text and redirect back to the payment step of the checkout process
+            Registry::getSession()->setVariable('payerror', '-102');
+            Registry::getSession()->setVariable('payerrortext', join('<br/>', $aErrorDescriptions));
+
+            $sRedirectUrl = Registry::getConfig()->getShopHomeUrl() . 'cl=payment';
```
Comment (ps-cgrach) [inline Extend/Controller/OrderController.php]: just to be sure, let's add the session id too.

Comment (ps-mstifter) [reply Extend/Controller/OrderController.php]: `Registry::getConfig()->getShopHomeUrl()` returns the shop base URL + index.php + the sid automatically.

File: Extend/Controller/OrderController.php [hunk 3]
```diff
@@ -232,22 +232,37 @@ private function _handleTransaction($oBasket, $oOrder)
...
+
+                // add error message and code to array for display on frontend
+                $aErrorDescriptions[] = $sDescription . ' (Error code: ' . $sCode . ')';
+
                 $oLogger->error("\t$sSeverity with code $sCode and message '$sDescription' occurred.");
             }
-            return;
+
+            // set the custom payment error code and text and redirect back to the payment step of the checkout process
+            Registry::getSession()->setVariable('payerror', '-102');
```
Comment (nickstamat) [inline Extend/Controller/OrderController.php]: Maybe we should assign this value to a constant at the top?

File: Extend/Controller/OrderController.php (hunk 2 above)
Comment (nickstamat) [reply Extend/Controller/OrderController.php]: If that's the case, don't you need to use an ampersand (`&`) before `cl=payment`, to join with the other params?

File: Extend/Controller/OrderController.php (hunk 3 above)
Comment (ps-cgrach) [reply Extend/Controller/OrderController.php]: 👍

File: Core/OrderHelper.php [hunk 4]
```diff
@@ -10,36 +10,39 @@
...
-use OxidEsales\Eshop\Core\Field;
-use OxidEsales\Eshop\Core\Registry;
 use OxidEsales\Eshop\Application\Model\Basket;
+use OxidEsales\Eshop\Application\Model\User;
 use OxidEsales\Eshop\Core\Exception\ArticleInputException;
 use OxidEsales\Eshop\Core\Exception\NoArticleException;
 use OxidEsales\Eshop\Core\Exception\OutOfStockException;
-use OxidEsales\Eshop\Application\Model\User;
-
+use OxidEsales\Eshop\Core\Field;
```
Comment (tomazpu) [inline Core/OrderHelper.php]: As you always slit up so nicely the uses please also group and separate them with a new line as in other files.

File: Core/OrderHelper.php [hunk 5]
```diff
@@ -146,26 +166,36 @@ public static function isFinalizeOrderSuccessful($iSuccess)
...
             $sDescription = $oStatus->getDescription();
+
+            // add error message and code to array for display on frontend
+            $aErrorDescriptions[] = $sDescription . ' (Error code: ' . $sCode . ')';
+
             $oLogger->error("\t$sSeverity with code $sCode and message '$sDescription' occurred.");
         }
+
+        // set the custom payment error code and text and redirect back to the payment step of the checkout process
+        Registry::getSession()->setVariable(self::PAY_ERROR_VARIABLE, self::PAY_ERROR_IDs);
```
[the remark this answers is not in the document]
Comment (ps-mstifter) [reply Core/OrderHelper.php]: 👍

File: Core/PaymentMethodHelper.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Returns a payment with the selected id.
+     *
+     * @param string $sPaymentId
+     *
+     * @return Payment
+     *
+     * @since 1.0.0
+     */
+    public static function getPaymentById($sPaymentId)
```
Comment (tomazpu) [inline Core/PaymentMethodHelper.php]: Great that you did this now, please also refactor the old code and use it everywhere you can. (NotifyHandler.php, PaymentGateway.php, Order.php, OrderController.php, PaymentMain.php)

Comment (ps-mstifter) [reply Core/PaymentMethodHelper.php]: It is now used in all the classes (thanks for the list btw 👍 )

**Pull request merged**